File: Haskell/Models/Line.hs
module Models.Line where

data Line = Line
  { originX :: Int,
    originY :: Int,
    endX :: Int
  }
  deriving (Show)

overlapsWith :: Line -> Line -> Bool
overlapsWith line anotherLine
  | not haveSameOriginY = False
  | originX line <= originX anotherLine = endX line >= originX anotherLine
  | otherwise = endX anotherLine >= originX line
  where
    haveSameOriginY = originY line == originY anotherLine

File: Prolog/models/line.pl
:- module(line, [create/4]).

create(OriginX, OriginY, EndX, Line):-
  Line = line(originX(OriginX), originY(OriginY), endX(EndX)).

originX(Line, OriginX):-
  Line = line(originX(OriginX), originY(_), endX(_)).

originY(Line, OriginY):-
  Line = line(originX(_), originY(OriginY), endX(_)).

endX(Line, EndX):-
  Line = line(originX(_), originY(_), endX(EndX)).
